File: frontend/src/lib/api/items.ts
import { auth } from '$lib/stores/auth.svelte';

export interface Item {
  id: string;
  workspace_id: string;
  title: string;
  content: string | null;
  item_type: string;
  metadata: Record<string, any>;
  tags: string[];
  user_id: string;
  created_at: string;
  updated_at: string;
}

export interface CreateItemPayload {
  workspace_id: string;
  title: string;
  content?: string | null;
  item_type: string;
  metadata?: Record<string, any>;
  tags?: string[];
}

export interface UpdateItemPayload {
  title?: string;
  content?: string | null;
  item_type?: string;
  metadata?: Record<string, any>;
  tags?: string[];
}

export interface ItemFilters {
  workspace_id?: string;
  item_type?: string;
  tags?: string[];
  search?: string;
  limit?: number;
  offset?: number;
}

function headers(): Record<string, string> {
  const h: Record<string, string> = { 'Content-Type': 'application/json' };
  if (auth.token) h.Authorization = `Bearer ${auth.token}`;
  return h;
}

async function handle<T>(res: Response): Promise<T> {
  if (!res.ok) {
    let msg = `Request failed: ${res.status}`;
    try {
      const err = await res.json();
      msg = err.error || err.message || msg;
    } catch {}
    throw new Error(msg);
  }
  if (res.status === 204) return undefined as T;
  return res.json();
}

function buildQuery(filters: ItemFilters): string {
  const params = new URLSearchParams();
  if (filters.workspace_id) params.set('workspace_id', filters.workspace_id);
  if (filters.item_type) params.set('item_type', filters.item_type);
  if (filters.tags && filters.tags.length > 0) params.set('tags', filters.tags.join(','));
  if (filters.search) params.set('search', filters.search);
  if (filters.limit !== undefined) params.set('limit', String(filters.limit));
  if (filters.offset !== undefined) params.set('offset', String(filters.offset));
  const qs = params.toString();
  return qs ? `?${qs}` : '';
}

export const itemsApi = {
  async list(filters: ItemFilters = {}): Promise<Item[]> {
    const res = await fetch(`/api/items${buildQuery(filters)}`, { headers: headers() });
    return handle<Item[]>(res);
  },

  async listByWorkspace(workspaceId: string, filters: Omit<ItemFilters, 'workspace_id'> = {}): Promise<Item[]> {
    const res = await fetch(`/api/workspaces/${workspaceId}/items${buildQuery(filters)}`, { headers: headers() });
    return handle<Item[]>(res);
  },

  async get(id: string): Promise<Item> {
    const res = await fetch(`/api/items/${id}`, { headers: headers() });
    return handle<Item>(res);
  },

  async create(payload: CreateItemPayload): Promise<Item> {
    const res = await fetch('/api/items', {
      method: 'POST',
      headers: headers(),
      body: JSON.stringify(payload)
    });
    return handle<Item>(res);
  },

  async update(id: string, payload: UpdateItemPayload): Promise<Item> {
    const res = await fetch(`/api/items/${id}`, {
      method: 'PATCH',
      headers: headers(),
      body: JSON.stringify(payload)
    });
    return handle<Item>(res);
  },

  async remove(id: string): Promise<void> {
    const res = await fetch(`/api/items/${id}`, {
      method: 'DELETE',
      headers: headers()
    });
    await handle<void>(res);
  }
};

File: backend/src/handlers/items.rs
use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    response::Json,
};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use crate::{AppState, db::schema::*};
use tracing::{error, info};
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize)]
pub struct CreateItemRequest {
    pub workspace_id: String,
    pub title: String,
    pub content: Option<String>,
    pub item_type: String,
    pub metadata: Option<Value>,
    pub tags: Option<Vec<String>>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct UpdateItemRequest {
    pub title: Option<String>,
    pub content: Option<String>,
    pub item_type: Option<String>,
    pub metadata: Option<Value>,
    pub tags: Option<Vec<String>>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ItemResponse {
    pub id: String,
    pub workspace_id: String,
    pub title: String,
    pub content: Option<String>,
    pub item_type: String,
    pub metadata: Value,
    pub tags: Vec<String>,
    pub user_id: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Deserialize)]
pub struct ItemFilters {
    pub workspace_id: Option<String>,
    pub item_type: Option<String>,
    pub tags: Option<String>, // Comma-separated tags
    pub search: Option<String>,
    pub limit: Option<u32>,
    pub offset: Option<u32>,
}

pub async fn create_item(
    State(state): State<AppState>,
    Json(payload): Json<CreateItemRequest>,
) -> Result<Json<ItemResponse>, StatusCode> {
    // TODO: Extract user_id from auth when auth is implemented
    let user_id = "temp_user_123";
    
    let metadata = payload.metadata.unwrap_or(serde_json::json!({}));
    let tags = payload.tags.unwrap_or_default();
    
    match state.db.create_item(
        user_id,
        &payload.workspace_id,
        &payload.title,
        payload.content.as_deref(),
        &payload.item_type,
        metadata,
        tags,
    ).await {
        Ok(item) => {
            info!("Item created: {}", item.id);
            Ok(Json(ItemResponse {
                id: item.id.to_string(),
                workspace_id: item.workspace_id,
                title: item.title,
                content: item.content,
                item_type: item.item_type,
                metadata: item.metadata,
                tags: item.tags,
                user_id: item.user_id,
                created_at: item.created_at.to_rfc3339(),
                updated_at: item.updated_at.to_rfc3339(),
            }))
        }
        Err(e) => {
            error!("Failed to create item: {}", e);
            Err(StatusCode::INTERNAL_SERVER_ERROR)
        }
    }
}

pub async fn get_items(
    State(state): State<AppState>,
    Query(filters): Query<ItemFilters>,
) -> Result<Json<Vec<ItemResponse>>, StatusCode> {
    // TODO: Extract user_id from auth when auth is implemented
    let user_id = "temp_user_123";
    
    let tags = filters.tags
        .map(|t| t.split(',').map(|s| s.trim().to_string()).collect::<Vec<String>>())
        .unwrap_or_default();
    
    match state.db.get_items_by_user(
        user_id,
        filters.workspace_id.as_deref(),
        filters.item_type.as_deref(),
        if tags.is_empty() { None } else { Some(tags) },
        filters.search.as_deref(),
        filters.limit.unwrap_or(50),
        filters.offset.unwrap_or(0),
    ).await {
        Ok(items) => {
            let response: Vec<ItemResponse> = items
                .into_iter()
                .map(|item| ItemResponse {
                    id: item.id.to_string(),
                    workspace_id: item.workspace_id,
                    title: item.title,
                    content: item.content,
                    item_type: item.item_type,
                    metadata: item.metadata,
                    tags: item.tags,
                    user_id: item.user_id,
                    created_at: item.created_at.to_rfc3339(),
                    updated_at: item.updated_at.to_rfc3339(),
                })
                .collect();
            Ok(Json(response))
        }
        Err(e) => {
            error!("Failed to get items: {}", e);
            Err(StatusCode::INTERNAL_SERVER_ERROR)
        }
    }
}

pub async fn get_item(
    State(state): State<AppState>,
    Path(item_id): Path<String>,
) -> Result<Json<ItemResponse>, StatusCode> {
    match state.db.get_item_by_id(&item_id).await {
        Ok(Some(item)) => {
            Ok(Json(ItemResponse {
                id: item.id.to_string(),
                workspace_id: item.workspace_id,
                title: item.title,
                content: item.content,
                item_type: item.item_type,
                metadata: item.metadata,
                tags: item.tags,
                user_id: item.user_id,
                created_at: item.created_at.to_rfc3339(),
                updated_at: item.updated_at.to_rfc3339(),
            }))
        }
        Ok(None) => Err(StatusCode::NOT_FOUND),
        Err(e) => {
            error!("Failed to get item: {}", e);
            Err(StatusCode::INTERNAL_SERVER_ERROR)
        }
    }
}

pub async fn update_item(
    State(state): State<AppState>,
    Path(item_id): Path<String>,
    Json(payload): Json<UpdateItemRequest>,
) -> Result<Json<ItemResponse>, StatusCode> {
    match state.db.update_item(
        &item_id,
        payload.title.as_deref(),
        payload.content.as_deref(),
        payload.item_type.as_deref(),
        payload.metadata,
        payload.tags,
    ).await {
        Ok(Some(item)) => {
            info!("Item updated: {}", item_id);
            Ok(Json(ItemResponse {
                id: item.id.to_string(),
                workspace_id: item.workspace_id,
                title: item.title,
                content: item.content,
                item_type: item.item_type,
                metadata: item.metadata,
                tags: item.tags,
                user_id: item.user_id,
                created_at: item.created_at.to_rfc3339(),
                updated_at: item.updated_at.to_rfc3339(),
            }))
        }
        Ok(None) => Err(StatusCode::NOT_FOUND),
        Err(e) => {
            error!("Failed to update item: {}", e);
            Err(StatusCode::INTERNAL_SERVER_ERROR)
        }
    }
}

pub async fn delete_item(
    State(state): State<AppState>,
    Path(item_id): Path<String>,
) -> Result<StatusCode, StatusCode> {
    match state.db.delete_item(&item_id).await {
        Ok(()) => {
            info!("Item deleted: {}", item_id);
            Ok(StatusCode::NO_CONTENT)
        }
        Err(e) => {
            error!("Failed to delete item: {}", e);
            Err(StatusCode::INTERNAL_SERVER_ERROR)
        }
    }
}

pub async fn get_items_by_workspace(
    State(state): State<AppState>,
    Path(workspace_id): Path<String>,
    Query(mut filters): Query<ItemFilters>,
) -> Result<Json<Vec<ItemResponse>>, StatusCode> {
    // Override workspace_id from path
    filters.workspace_id = Some(workspace_id);
    get_items(State(state), Query(filters)).await
}